### PR TITLE
Convert to protogen Plugin and protojson instead of jsonpb

### LINKED ISSUE #6: Move to google.golang.org/protobuf/encoding/protojson
With `github.com/golang/protobuf/jsonpb` being deprecated this repository should probably instead (or optionally) target `google.golang.org/protobuf/encoding/protojson`.

### PR DESCRIPTION
**Pull request by chkohner**
This updates the plugin to use the [protogen framework](https://pkg.go.dev/google.golang.org/protobuf/compiler/protogen) introduced in [protobuf api V2](https://blog.golang.org/protobuf-apiv2). It removes a lot of redundant logic and also updates the templates to use [`encoding/protojson`](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson) instead of `jsonpb` (which is now deprecated).

It now no longer depends on the grpc-gateway ecosystem (which looks like it was moved to an internal package anyway).

Fixes #6

### CODE REVIEW

**Comment by mitchellh** [comment]
Yes! Awesome. I'll keep an eye on this and eventuallly test locally when you're ready. I've been wanting to do this, thank you for taking it on.

**Comment by chkohner** [comment]
My initial testing seems to work fine, and I think I caught all the necessary README updates. The biggest difference is that you have to explicitly give it a value (e.g. `enums_as_ints=true`) as the protogen plugin doesn't parse boolean flags quite the same way.

Please test locally and let me know!

**Comment by glerchundi** [comment]
Wow, good job! @chkohner!

File: gen/template.go [hunk 1]
```diff
@@ -1,93 +1,94 @@
...
 }
 
 // UnmarshalJSON implements json.Unmarshaler
-func (msg *{{.TypeName}}) UnmarshalJSON(b []byte) error {
-	return (&jsonpb.Unmarshaler{
-	  AllowUnknownFields: {{.AllowUnknownFields}},
-	}).Unmarshal(bytes.NewReader(b), msg)
+func (msg *{{.GoIdent.GoName}}) UnmarshalJSON(b []byte) error {
+	return protojson.UnmarshalOptions {
+		DiscardUnknown: {{.AllowUnknownFields}},
```
Comment (bluestealth) [inline gen/template.go:90]: I think this should be `{{not .AllowUnknownFields}},`

Comment (chkohner) [reply gen/template.go:90]: According to the protojson docs:

> 	// If DiscardUnknown is set, unknown fields are ignored.
	DiscardUnknown bool

The field controlling the functionality used to be called `AllowUnknownFields` in jsonpb.

> 	// AllowUnknownFields specifies whether to allow messages to contain
	// unknown JSON fields, as opposed to failing to unmarshal.
	AllowUnknownFields bool

But the [behavior is the same](https://github.com/golang/protobuf/blob/4846b58453b3708320bdb524f25cc5a1d9cda4d4/jsonpb/decode.go#L118-L121).

Comment (bluestealth) [reply gen/template.go:90]: Confusing, but makes sense. Thanks for clarifying.

Comment (chkohner) [reply gen/template.go:90]: Yeah, I'm not sure why they changed the field name. I prefer the old one myself. 😄

**Comment by lopezator** [comment]
I would also like to get this merged! 🚀 . Nicely done @chkohner

**Comment by hienphamlabs** [comment]
Nice work @chkohner.
I used @chkohner 's branch in productions and it works nicely.
Also I update some code that work our usecase:  that explicit define `go_package`, format proto file using clang-format and bumps protobuf to 1.26.0. (https://github.com/hienduyph/protoc-gen-go-json/commit/838943ed7194cc821086214e27566a27bd31b468)

**Comment by elvizlai** [comment]
Maybe we should also support: `generator protoc-gen-go-json hasn't been updated to support optional fields in proto3`

**Pull request merged**